### DISCUSSION TITLE
OpenCart autocomplete code

### DISCUSSION THREAD

**Discussion by rsuchovsky** [opencart/opencart#12733] [Q&A]
Hello, for many years I have been looking for a way to automatically complete method names in opencart.

**The only thing I could find is this for OC3**, but it doesn't solve all the methods in catalog, account, etc..
https://github.com/SevenPowerX-PHP/opencart-phpstorm

What I'm looking for is a way to automatically complete the method name. Also get to the method using **CTRL + LEFT CLICK** if I **click on the method name**. It currently tells me that it cannot find the declared method.
### When writing code it doesn't offer method name hint etc... **it's very frustrating** 😵

![ide](https://github.com/opencart/opencart/assets/29576358/bee828a5-4104-4d91-8255-abb0d4306e7a)

How to arrange this in **PhpStorm**, is such a thing possible in OpenCart code?
Or are you all programming in opencart without such options?

### I will be very grateful for any answer, it is very bad if the code does not offer an automatic hint of the name of declared methods, functions, class ..

**Comment by mhcwebdesign** [comment]
This is not an OpenCart issue, rather, it's about the text editor you use. I use Kate which does what I need, including name suggestions from what it found within a PHP file.

**Comment by stalker780** [comment]
@mhcwebdesign do Kate show autocomplete functions list if you type in any controller `$this->cart->` ?
Or for `$this->model_catalog_product->` ?


@rsuchovsky 
I suppose there is no IDE which can work correctly with OC structure. Cause it is based on variables.

I have tested this solution https://github.com/mithereal/opencart-ide-autocomplete for OC2.
It scans your store and adds comments into system/engine/controller.php
At least basic libraries and models support autocomplete.

I remember I have made a modified version for OC3. But don't think I can find it now.

**Comment by rsuchovsky** [comment]
@mhcwebdesign I tried Kate now, but it didn't work as well as I imagined. But thanks for the tip.

@stalker780 For OC3, this one is great:
https://github.com/lSanik/opencart_phpstorm_autocomplete
Works well:
![ide-oc3](https://github.com/opencart/opencart/assets/29576358/a333395e-de25-48c8-8eeb-c3858a1d3e77)

### It would be great to have that option in OC4 as well. I couldn't find anything for OC4.

**Comment by stalker780** [comment]
@rsuchovsky This method works in OC4 too.
But you need to set correct namespaces according to OC4 structure.

I have manually modified system/engine/controller.php adding before `class Controller {`
```
     /**
	* @property string $id
	* @property string $template
	* @property array $children
	* @property array $data
	* @property string $output
	* @property \Opencart\System\Engine\Loader $load
	* @property \Opencart\System\Library\Cart\Cart $cart
	* @property \Opencart\System\Library\Cart\User $user
	* @property \Opencart\System\Library\Cart\Customer $customer
	* @property \Opencart\System\Library\Mail $mail
	* @property \Opencart\System\Library\Encryption $encryption
	* @property \Opencart\System\Engine\Event $event
	* @property \Opencart\System\Library\Db $db
	* @property \Opencart\System\Library\Url $url
	* @property \Opencart\System\Library\Request $request
	* @property \Opencart\System\Library\Response $response
	* @property \Opencart\System\Library\Cache $cache
	* @property \Opencart\System\Library\Session $session
	* @property \Opencart\System\Library\Language $language
	* @property \Opencart\System\Library\Document $document
	* @property \Opencart\Catalog\Model\Catalog\Product $model_catalog_product1
	* @property \Opencart\Admin\Model\Catalog\Product $model_catalog_product
	**/
```

and this works. So this script can be easily adapted for OC4.

Now I recall why I played with it, but did not use it. Because both catalog and admin controllers are childs of Opencart\System\Engine\Controller class. You can use only one of model_catalog_product. And you can do nothing about it without changing OC structure.
If you choose catalog, it will show you catalog methods in admin. 
If you choose admin, it will show you admin methods in catalog like addProduct(). 
It is confusing.

Nevertheless it works correctly for libraries in most IDEs.

![image](https://github.com/opencart/opencart/assets/6297070/2f50a15b-0df9-4f04-a7bd-542418e34d42)

**Comment by rsuchovsky** [comment]
@stalker780  Thanks 🙏

**This saves me a lot of time writing code.**
### It would be good to have such a script officially available for download.
_Although it is not perfect. I think not one developer would appreciate it_
